File: leetcode/413-arithmetic-slices/main.go
package main

/*
	naive approach
	1. for each num, look for forward to find the arithemetic boundary
	2. calculate the count of the arithemetic intermediate array
	Time    O(n^2)
	Space   O(1)
	not gonna implement
*/

/*
	1st appraoch
	- math, use n*(n+1)/2 to calculate the combination count instead of doing brute force
	Time    O(n)
	Space   O(1)
	20ms beats 100%
	31jan2019
*/
func numberOfArithmeticSlices(A []int) int {
	if len(A) < 3 {
		return 0
	}
	diff := A[1] - A[0]
	res := 0
	start := 0
	for i := 1; i < len(A); i++ {
		if A[i]-A[i-1] == diff {
			if i+1 == len(A) {
				res += calCnt(start, i)
			}
			continue
		}
		res += calCnt(start, i-1)
		// be careful: set the next start point from the previous item
		// take this case into consideration [1, 2, 3, 8, 13]
		start = i - 1
		diff = A[i] - A[i-1]
	}
	return res
}

func calCnt(start, i int) int {
	n := i - start + 1
	if n >= 3 {
		cnt := n - 2
		return cnt * (cnt + 1) / 2
	}
	return 0
}

func main() {

}

File: leetcode/413-arithmetic-slices/main.py
"""
    1st: math

    e.g. [1, 2, 3, 4, 5, 7, 9, 11]
    diff =  [1, 1, 1, 1, 2, 2, 2]

    for [1, 2, 3, 4, 5], we can come up with 4(3)/2 = 6
    for [5, 7, 9, 11], we can come up with 3(2)/2 = 3
    result = 6 + 3 = 9

    Time    O(N)
    Space   O(N)
    28 ms, beats 97.75%
"""


class Solution:
    def numberOfArithmeticSlices(self, A: List[int]) -> int:
        if len(A) < 3:
            return 0
        n = len(A)
        diffs = []
        for i in range(1, n):
            diffs.append(A[i] - A[i-1])
        res = 0
        cur = diffs[0]
        count = 1
        for i in range(1, len(diffs)):
            if diffs[i] == cur:
                count += 1
            else:
                if count >= 2:
                    res += self.cal(count)
                cur = diffs[i]
                count = 1
        if count >= 2:
            res += self.cal(count)
        return res

    def cal(self, n):
        return n * (n-1) // 2
